### ISSUE TITLE
GPU performance tuning

### ISSUE THREAD

**Issue by stanos4** [atomistic-machine-learning/schnetpack#199]
Hello,

I'm looking for performance advice. I try to reproduce your results with QM9 database. So if I understand correctly, this should reproduce the results and performance as published at your paper "SchNetPack: A Deep Learning Toolbox For Atomistic Systems" (https://arxiv.org/pdf/1809.01072v1.pdf):

     spk_run.py train schnet qm9 ./qm9.db ./ --split 109000 1000 --cuda --batch_size=100 --lr_decay=0.5

However, my timing is around 50s/epoch vs <6s in your case (Fig. 4). 
system: Xeon E5-2660v4 at ~3GHz; Tesla P100 (16GB) with CUDA 10.1.105; Conda installation of Numpy and several other things and pip installation of the Schnetpack (schnetpack-0.3-py3-none-any.whl)
nvidia-smi reports <~80% GPU-Util

I also tried system with V100 GPUs, but the reported GPU utilization goes down to ~50%, with time/epoch going down a bit (maybe 30-50s)
I run directly from memory (to avoid access to HDD).

Do you have any suggestions for improving the performance closer to what you got? Or any ideas to narrow down possible root-causes?
(p.s., it may be a problem with my system/installation, as this is an offline system where packages have to be downloaded manually and so some things can get tricky/messy with python)
Thank you!

**Comment by stanos4** [comment]
here is the args.json content

    {
        "batch_size": 100,
        "checkpoint_interval": 1,
        "cuda": true,
        "cutoff": 10.0,
        "cutoff_function": "cosine",
        "datapath": "./qm9.db",
        "dataset": "qm9",
        "environment_provider": "simple",
        "features": 128,
        "interactions": 6,
        "keep_n_checkpoints": 3,
        "log_every_n_epochs": 1,
        "logger": "csv",
        "lr": 0.0001,
        "lr_decay": 0.5,
        "lr_min": 1e-06,
        "lr_patience": 25,
        "max_epochs": 5000,
        "max_steps": null,
        "mode": "train",
        "model": "schnet",
        "modelpath": "./",
        "n_epochs": 1000,
        "num_gaussians": 50,
        "overwrite": false,
        "parallel": false,
        "property": "energy_U0",
        "remove_uncharacterized": false,
        "seed": null,
        "split": [
            109000,
            1000
        ],
        "split_path": null
    }

**Comment by stanos4** [comment]
So I have tested another system (Google cloud platform), Debian 10 (4.9.0-11-amd64) with pre-installed Cuda 10.1 (driver v418.87.01), Conda 4.8.0 and Pytorch 1.3, on K80 GPU and the results are in line with my previous timings:

    pip install schnetpack --user
    ~/.local/bin/spk_run.py train schnet qm9 ./database_stan.db ./ --split 109000 1000 --cuda --batch_size=128 --lr_decay=0.5
    
    Time,Learning rate,Train loss,Validation loss,MAE_energy_U0,RMSE_energy_U0
    128.80586457252502, 0.0001, 0.7087687910171824, 0.2034348258972168, 0.31339501953125, 0.45103749943570853
    256.7535901069641, 0.0001, 0.14077266636299432, 0.130882999420166, 0.25920751953125, 0.36177755516361987
    385.08452439308167, 0.0001, 0.09236586321822, 0.09484081429243088, 0.2081865234375, 0.30796235749971096


and

    ~/.local/bin/spk_run.py train schnet qm9 ./database_stan.db ./ --split 109000 1000 --cuda --batch_size=100 --lr_decay=0.5
    
    Time,Learning rate,Train loss,Validation loss,MAE_energy_U0,RMSE_energy_U0
    129.34601593017578, 0.0001, 0.513197237909387, 0.14928167536854745, 0.26595458984375, 0.3863698735065175
    258.76035237312317, 0.0001, 0.10632533788202553, 0.12618914023041725, 0.26971044921875, 0.3552311123491961
    387.8785915374756, 0.0001, 0.07767599711724378, 0.07654303908348084, 0.19070361328125, 0.27666412900044296


GPU utilization as reported by nvidia-smi is ~90% at both cases. However, only 4 CPU cores are utilized from the 8 available (from "top", the used cores are running at ~<50%, so this shouldn't be probably a problem, but still, are the 4 cores for the GPU version "hard-coded" somewhere?). Therefore, am I doing something completely different or is the publication (arxiv version) graph missing 1 zero?

Thank you for any suggestions.

**Comment by Stefaanhess** [comment]
Hey @stanos4 ,
sorry for the late reply.
It seems like there is an issue with the figure in the paper. A minute per epoch should be a reasonable time for training.

Best regards,
Stefaan

**Comment by stanos4** [comment]
Thank you!
Best,
Stan.

**Comment by Stefaanhess** [comment]
Hey @stanos4 ,

I just saw, that I did not answer your second question:
>are the 4 cores for the GPU version "hard-coded" somewhere?

You can set the number of cpu workers in the dataloader. It is hard-coded to 4:

https://github.com/atomistic-machine-learning/schnetpack/blob/d45ae7523f3232bfc349ec25e9f9b3f26d14358a/src/schnetpack/utils/script_utils/data.py#L89

Best,
Stefaan

**Comment by stanos4** [comment]
Thank you! I couldn't see it before with grepping number 4 through the source :)
